README.md
# Filename-Manager


[![pypi](https://img.shields.io/pypi/v/FilenameManager.svg)](https://pypi.org/project/FilenameManager/)
[![python](https://img.shields.io/pypi/pyversions/FilenameManager.svg)](https://pypi.org/project/FilenameManager/)
[![Build Status](https://github.com/cemde/FilenameManager/actions/workflows/dev.yml/badge.svg)](https://github.com/cemde/FilenameManager/actions/workflows/dev.yml)
[![codecov](https://codecov.io/gh/cemde/FilenameManager/branch/main/graphs/badge.svg)](https://codecov.io/github/cemde/FilenameManager)



A simple Python library to handle filenames automatically for artifacts.


[//]: <> (* Documentation: <https://cemde.github.io/FilenameManager>)
* GitHub: <https://github.com/cemde/FilenameManager>
* PyPI: <https://pypi.org/project/FilenameManager/>

## Install

```bash
pip install filenamemanager
```

## Features

This library helps to convert a set of parameters into filenames and vice versa.

When running a script depending on parameters `par1` (str) and `par2` (float), you might want to save artifacts with a filename generated by these parameters.

You can decide how to format each parameter in the saved filename, e.g. `fp4.4` will padd the float with leading `0`s and with 4 digits. For example, `1.2` will be `0001.2000`.

### Save artifacts
```python
from filename_manager import FilenameManager

fm = FilenameManager({"par1": "str", "par2": "fp.4.4}"}, prefix="subfolder/prefix", postfix=".csv")

# define your parameters
par1 = "Hello"
par2 = "3.11"

# super fancy script
results = "placeholder"

# create filename
filename=fm.encode(par1=par1, par2=par2)

# save
...

```

the filename will be `subfolder/prefix_par1_Hello_par2_0003.1100.csv`.

### Load artifacts

```python
from filename_manager import FilenameManager

fm = FilenameManager({"par1": "str", "par2": "fp.4.4}"}, prefix="subfolder/prefix", postfix=".csv")

# get your filename
filename = "subfolder/prefix_par1_Hello_par2_0003.1100.csv"

# parse filename
pars = fm.decode(filename)
```

The variable `pars` is a dictionary with `{"par1": "Hello", "par2": 3.11}`.


## Documentation

The documentation is given in the docstrings. Only two classes exist. `Parameter` and `FilenameManager`.

The supported formats are

- `fp*.*` for floating point parameters. E.g. "fp4.3" formats 1.2 as "0001.200"
- `str` for string parameters.
- `str*` for string parameters with * characters padded with 0 in the end.
- `*str` for string parameters with * characters padded with 0 in the beginning.
- `bool` for boolean parameters.
- `*int` for integer parameters with * leading zeros.

## Credits

This package was created with [Cookiecutter](https://github.com/audreyr/cookiecutter) and the [waynerv/cookiecutter-pypackage](https://github.com/waynerv/cookiecutter-pypackage) project template.


CHANGELOG.md
# Changelog

## 0.3.0 (2023-04-06)

* Fixed unneeded and restrictive dependency on jinja2.

## 0.2.0 (2022-10-11)

* BUGFIX: Fixed bug to prevent decoding with negative floats.

## 0.1.0 (2022-08-27)

* First release on PyPI.


tests/__init__.py
"""Unit test package for FilenameManager."""


tests/test_parameter.py
from filename_manager import Parameter, parameter


def test_attributes():
    par = Parameter("par1", "bool")
    assert par.name == "par1"
    assert par.format == "bool"


def test_float():
    par = Parameter("par", "fp2.3")
    p_e = par.encode(133.234113)
    assert p_e == "133.234", f"Encoded {p_e} != 133.234"
    p_e = par.encode(1)
    assert p_e == "01.000", f"Encoded {p_e} != 01.000"
    assert par.decode(par.encode(133.234113)) == 133.234
    assert par.decode(par.encode(1.0)) == 1.0

    par = Parameter("par", "fp1.1")
    p_e = par.encode(133.234113)
    assert p_e == "133.2", f"Encoded {p_e} != 133.2"
    p_e = par.encode(1)
    assert p_e == "1.0", f"Encoded {p_e} != 1.0"
    assert par.decode(par.encode(133.354113)) == 133.4
    assert par.decode(par.encode(1.0)) == 1.0

    par = Parameter("par", "fp2.3")
    par = par.decode("-00.900")
    assert par == -0.9


def test_int():
    par = Parameter("par", "int")
    assert par.encode(133) == "133"
    assert par.encode(-100) == "-100"

    assert par.decode("-100") == -100
    assert par.decode("133") == 133

    par = Parameter("par", "4int")
    assert par.encode(133) == "0133"
    assert par.encode(-1) == "-001"
    assert par.encode(123456) == "123456"

    assert par.decode("-0100") == -100
    assert par.decode("0133") == 133
    assert par.decode("123456") == 123456


def test_str():
    par = Parameter("par", "str")
    assert par.encode("Blub45") == "Blub45"
    assert par.decode("Blub45") == "Blub45"
    assert par.decode("45.3") == "45.3"

    par = Parameter("par", "8str")
    assert par.encode("Blub45") == "00Blub45"
    assert par.encode("Blub45Blub12") == "Blub45Blub12"
    assert par.decode("00Blub45") == "Blub45"
    assert par.decode("000045.3") == "45.3"
    assert par.decode("Blub45Blub12") == "Blub45Blub12"

    par = Parameter("par", "str8")
    assert par.encode("Blub45") == "Blub4500"
    assert par.encode("Blub45Blub12") == "Blub45Blub12"
    assert par.decode("Blub4500") == "Blub45"
    assert par.decode("45.30000") == "45.3"
    assert par.decode("Blub45Blub12") == "Blub45Blub12"


def test_bool():
    par = Parameter("par", "bool")
    assert par.encode(True) == "True"
    assert par.decode("True") == True
    assert par.encode(False) == "False"
    assert par.decode("False") == False


tests/test_filename_manager.py
#!/usr/bin/env python
"""Tests for `filename_manager` package."""

from filename_manager import FilenameManager, Parameter
import pytest


def get_pars():
    return [
        Parameter("s", "fp2.3"),
        Parameter("m", "8str"),
        Parameter("t_t_q", "str"),
        Parameter("k", "str10"),
        Parameter("p_q_0", "int"),
        Parameter("g", "4int"),
        Parameter("v", "fp1.1"),
        Parameter("b", "bool"),
    ]


def test_with_parameter_list():
    fm = FilenameManager(parameters=get_pars(), postfix='.csv')
    assert (
        fm.encode(s=2.345, m="abc", t_t_q="de", k="blob", p_q_0=54321, g=12, v=1.1, b=False)
        == 's_02.345_m_00000abc_t_t_q_de_k_blob000000_p_q_0_54321_g_0012_v_1.1_b_False.csv'
    )
    assert fm.decode('s_02.345_m_00000abc_t_t_q_de_k_blob000000_p_q_0_54321_g_0012_v_1.1_b_False.csv') == {
        "s": 2.345,
        "m": "abc",
        "t_t_q": "de",
        "k": "blob",
        "p_q_0": 54321,
        "g": 12,
        "v": 1.1,
        "b": False,
    }


def test_with_str_dict():
    fm = FilenameManager({"sigma": "fp2.3", "method": "str", "phi": "int", "v": "fp1.1"}, postfix='.csv')
    assert fm.encode(sigma=1.246, method="tata_ta", phi=15, v=1.0) == 'sigma_01.246_method_tata_ta_phi_15_v_1.0.csv'
    assert fm.decode('sigma_01.200_method_tata_ta_phi_5_v_1.3.csv') == {
        'sigma': 1.2,
        'method': 'tata_ta',
        'phi': 5,
        'v': 1.3,
    }


@pytest.mark.xfail
def test_encode_missing_parameter():
    pass


@pytest.mark.xfail
def test_encode_too_many_parameters():
    pass


def test_more_decoding():
    fm = FilenameManager(
        {'n_data': '6int', 'n_bins': '3int', 'conf': 'fp1.2', 'c': 'fp1.2', 'n_sim': '9int'},
        prefix='pkls/',
        postfix='.pickle',
    )
    filename = "pkls/n_data_000100_n_bins_001_conf_-0.50_c_0.40_n_sim_001000000.pickle"
    pars = fm.decode(filename)
    assert pars["conf"] == -0.5


filename_manager/filename_manager.py
"""Main module."""

import re
from typing import Any, Dict, List, Union

from .parameter import Parameter


class FilenameManager:
    """The FilenameManager generates and reads filenames automatically.

    Attributes:
        parameters (List[Parameter]): List of parameters.
        prefix (str): Prefix of the filename.
        postfix (str): Postfix of the filename.
        pattern (str): Regex pattern of the filename.
    """

    def __init__(self, parameters: Union[List[Parameter], Dict[str, str]], prefix: str = '', postfix: str = ''):
        """This will encode and decode filenames composed of parameters.

        Specify the format of each parameter and generate a file name.
            "fp*.*" for floating point parameters. E.g. "fp4.3" formats 1.2 as "0001.200"
            "str" for string parameters.
            "str*" for string parameters with * characters padded with 0 in the end.
            "*str" for string parameters with * characters padded with 0 in the beginning.
            "bool" for boolean parameters.
            "*int" for integer parameters with * leading zeros.

        Args:
            parameters (Union[List[Parameter], Dict[str, str]]): specifies the parameters with names and formats.
            prefix (str, optional): Prefix, e.g. path to file. Defaults to ''.
            postfix (str, optional): Postfix, e.g. file type. Defaults to ''.
        """
        if isinstance(parameters, dict):
            parameters = [Parameter(n, f) for n, f in parameters.items()]
        self.parameters: List[Parameter] = parameters

        pattern = [f"{p.name}_({p.pattern})" for p in parameters]
        self.prefix, self.postfix = prefix, postfix
        self.pattern = prefix + '_'.join(pattern) + postfix

    def encode(self, **kwargs) -> str:
        """Convert parameters to a file name.

        Args:
            **kwargs: parameters to be encoded.

        Returns:
            str: file name.
        """
        # check if kwargs is missing parameters
        missing_kwargs = [k.name for k in self.parameters if k.name not in kwargs]
        if len(missing_kwargs) > 0:
            raise ValueError(f"Missing parameters: {missing_kwargs}")

        # check if kwargs contains extra parameters not known
        missing_pars = [p for p in kwargs if p not in [k.name for k in self.parameters]]
        if len(missing_pars) > 0:
            raise ValueError(f"Unknown parameters: {missing_pars}")

        s = [f"{p.name}_{p.encode(kwargs[p.name])}" for p in self.parameters]
        return self.prefix + '_'.join(s) + self.postfix

    def decode(self, string: str) -> Dict[str, Any]:
        """Convert file name to parameters.

        Args:
            string (str): File name string.

        Returns:
            Dict[str, Any]: Parameters.
        """
        m = re.match(self.pattern, string)
        if m is None:
            raise ValueError(f"Filename {string} does not match pattern {self.pattern}")
        args = {}
        for i, p in enumerate(self.parameters):
            args[p.name] = p.decode(m.group(i + 1))
        return args


filename_manager/parameter.py
"""This module defines the Parameter class."""
import re
from typing import Any


class Parameter:
    """The Parameter class encodes and decodes individual parameters to and from strings.

    Attributes:
        name (str): The name of the parameter.
        format (str): The format.
        pattern (str): The regular expression pattern.
    """

    def __init__(self, name: str, format: str) -> None:
        """Create a parameter class.

        This parameter can encode and decode values to and from strings. Specify
        the format of each parameter and generate a file name.
            "fp*.*" for floating point parameters. E.g. "fp4.3" formats 1.2 as "0001.200"
            "str" for string parameters.
            "str*" for string parameters with * characters padded with 0 in the end.
            "*str" for string parameters with * characters padded with 0 in the beginning.
            "bool" for boolean parameters.
            "*int" for integer parameters with * leading zeros.

        Args:
            name (str): The name of the parameter.
            format (str): The format.
        """
        self.name = name
        self.format = format

        if "fp" in format:
            m = re.match(r'fp(\d+).(\d+)', format)
            assert m is not None, f"Invalid format: {format}"
            g1, g2 = int(m.group(1)), int(m.group(2))
            self.pattern = fr'-?\d{{{g1}}}.\d{{{g2}}}'
            self._decode_fn = float
            self._encode_fn = lambda x: f"{x:0{g1+g2+1}.{g2}f}"

        elif "bool" in format:
            self.pattern = 'True|False'
            self._decode_fn = lambda x: x == "True"  # type: ignore
            self._encode_fn = lambda x: str(x)

        elif "int" in format:
            if "int" == format:
                self.pattern = r'\d+'
                self._decode_fn = int  # type: ignore
                self._encode_fn = lambda x: str(x)
            else:
                n_digits = format[0]
                self.pattern = fr'\d{{{n_digits}}}'
                self._decode_fn = int  # type: ignore
                self._encode_fn = lambda x: f"{x:0{n_digits}d}"

        elif "str" in format:
            if format == "str":
                self.pattern = r'\w+'
                self._decode_fn = str  # type: ignore
                self._encode_fn = lambda x: x

            else:
                n_char = int(re.findall(r'\d+', format)[0])
                self.pattern = fr'\w{{{n_char}}}'
                self._decode_fn = lambda x: x.replace("0", "")  # type: ignore
                pad_at_back = format.startswith("str")

                def encfn(x):
                    pad_str = "0" * (n_char - len(x))
                    if pad_at_back:
                        return x + pad_str
                    else:
                        return pad_str + x

                self._encode_fn = encfn

    def encode(self, val: Any) -> str:
        """Convert parameters to a file name.

        Args:
            val (Any): value to be encoded.

        Returns:
            str: encoded value.
        """
        return self._encode_fn(val)

    def decode(self, val: str) -> Any:
        """Convert string to raw value.

        Args:
            val (str): Input string.

        Returns:
            Any: Decoded raw value.
        """
        return self._decode_fn(val)
